### PR TITLE
FIX: Ensure disabling 2FA works as expected

### PR DESCRIPTION
**Pull request by tshenry**
This will fix the bug reported [here on Meta](https://meta.discourse.org/t/cant-disable-2fa/160500)

The second factor management page has previously included a "Disable" button. Selecting this button brings up a confirmation:

old

Instead of disabling all registered TOTP Authenticators **and** Security Keys, confirming only disables all TOTP Authenticators.

---

This PR will change the "Disable" button on the second factor management page to say "Disable All," and make a few styling/copy improvements to the confirmation:

Screen Shot 2020-08-13 at 6 45 51 PM

Screen Shot 2020-08-19 at 3 31 25 PM

It also fixes the behavior so that **all** second factor methods are removed when using the "Disable All" button.

**Notes**

- The test was tricky and takes heavy inspirations from other tests. I imagine there may be a more elegant or complete implementation, but I'm pretty sure I have the most important functionality tested. If I need to make any changes, let me know.

- Both second factor methods can be disabled individually instead of using the "Disable All" button using the pencil icon buttons showing in the above screenshot. What I came to realize is that the implementation for individual disabling is quite different than the disable all method. The "Disable All" button deletes the relevant `user_second_factors` and `user_security_keys` records from the database, whereas disabling the second factors individually sets `enabled: false` in the relevant record and keeps it around. This is not new behavior and nothing is technically broken, I just wanted to bring it up in case we might want to have consistency. If anything needs to change, I imagine it can be addressed in a follow-up commit.

### CODE REVIEW

File: spec/requests/users_controller_spec.rb [hunk 1]
```diff
@@ -4080,6 +4080,43 @@ def create_totp
...
+
+        it 'should disable all totp and security keys' do
+          expect do
+            put "/u/disable_second_factor.json"
+          end.to change { Jobs::CriticalUserEmail.jobs.length }.by(1)
+
+          expect(response.status).to eq(200)
+
+          expect(user.reload.user_second_factors).to be_empty
+          expect(user.reload.security_keys).to be_empty
```
Comment (tgxworld) [inline spec/requests/users_controller_spec.rb:4109]:
```suggestion
          expect(user.reload.user_second_factors).to be_empty
          expect(user.security_keys).to be_empty
```

File: spec/requests/users_controller_spec.rb [hunk 2]
```diff
@@ -4080,6 +4080,43 @@ def create_totp
...
+      context 'when user has a registered totp and security key' do
+        before do
+          expect(user.reload.user_second_factors).to be_empty
+          expect(user.reload.security_keys).to be_empty
+
+          totp_second_factor = Fabricate(:user_second_factor_totp, user: user)
+          security_key_second_factor = Fabricate(:user_security_key, user: user, factor_type: UserSecurityKey.factor_types[:second_factor])
+
+          expect(user.reload.user_second_factors.totps.first).to eq(totp_second_factor)
+          expect(user.reload.security_keys.first).to eq(security_key_second_factor)
```
Comment (tgxworld) [inline spec/requests/users_controller_spec.rb:4098]:
```suggestion
          expect(user.reload.user_second_factors.totps.first).to eq(totp_second_factor)
          expect(user.security_keys.first).to eq(security_key_second_factor)
```

File: spec/requests/users_controller_spec.rb [hunk 3]
```diff
@@ -4080,6 +4080,43 @@ def create_totp
...
+  describe '#disable_second_factor' do
+    context 'when logged in with secure session' do
+      before do
+        sign_in(user)
+        stub_secure_session_confirmed
+      end
+      context 'when user has a registered totp and security key' do
+        before do
+          expect(user.reload.user_second_factors).to be_empty
+          expect(user.reload.security_keys).to be_empty
```
Comment (tgxworld) [inline spec/requests/users_controller_spec.rb:4092]:
```suggestion
          expect(user.user_second_factors).to be_empty
          expect(user.security_keys).to be_empty
```

File: spec/requests/users_controller_spec.rb [hunk 4]
```diff
@@ -4080,6 +4080,43 @@ def create_totp
...
+          expect do
+            put "/u/disable_second_factor.json"
+          end.to change { Jobs::CriticalUserEmail.jobs.length }.by(1)
+
+          expect(response.status).to eq(200)
+
+          expect(user.reload.user_second_factors).to be_empty
+          expect(user.reload.security_keys).to be_empty
+
+          job_args = Jobs::CriticalUserEmail.jobs.first["args"].first
```
Comment (tgxworld) [inline spec/requests/users_controller_spec.rb:4111]: Let's use the helpers in https://github.com/discourse/discourse/blob/0967ce478d85ad5053522944b36ed9212f867457/spec/support/sidekiq_helpers.rb#L19

File: app/assets/javascripts/discourse/app/controllers/preferences/second-factor.js [hunk 5]
```diff
@@ -119,12 +120,17 @@ export default Controller.extend(CanCheckEmails, {
...
-          if (result) {
+      const message = I18n.t("user.second_factor.disable_confirm");
+      const buttons = [
+        {
+          label: I18n.t("cancel"),
+          class: "d-modal-cancel",
+          link: true
+        },
+        {
+          label: `${iconHTML("ban")}` + I18n.t("user.second_factor.disable"),
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/controllers/preferences/second-factor.js:131]:
```suggestion
          label: `${iconHTML("ban")}${I18n.t("user.second_factor.disable")}`
```

File: spec/requests/users_controller_spec.rb [hunk 6]
```diff
@@ -4080,6 +4080,38 @@ def create_totp
...
+    context 'when logged in with secure session' do
+      before do
+        sign_in(user)
+        stub_secure_session_confirmed
+      end
+      context 'when user has a registered totp and security key' do
+        before do
+          expect(user.user_second_factors).to be_empty
+          expect(user.security_keys).to be_empty
+
```
Comment (tgxworld) [inline spec/requests/users_controller_spec.rb:4093]:
```suggestion
```

All tests start from a clean slate so we don't need to check for this.

File: spec/requests/users_controller_spec.rb [hunk 7]
```diff
@@ -4080,6 +4080,38 @@ def create_totp
...
+          totp_second_factor = Fabricate(:user_second_factor_totp, user: user)
+          security_key_second_factor = Fabricate(:user_security_key, user: user, factor_type: UserSecurityKey.factor_types[:second_factor])
+
+          expect(user.reload.user_second_factors.totps.first).to eq(totp_second_factor)
+          expect(user.security_keys.first).to eq(security_key_second_factor)
```
Comment (tgxworld) [inline spec/requests/users_controller_spec.rb:4098]:
```suggestion
```

The use of the fabricator ensures that the object is successfully created so we don't need these assertions too.

**Comment by tgxworld** [comment]
@tshenry Just to confirm, do we need to disable backup codes too?

**Comment by tshenry** [comment]
>do we need to disable backup codes too?

Good question, @tgxworld! I can't think of any reason why backup codes should hang around after someone uses the "Delete All" option, so I'll ensure they are destroyed as well.

I really appreciate all of the feedback and explanations surrounding the tests. I'm still getting the hang of writing them, so everything you've mentioned is extremely helpful ❤️

**Comment by tshenry** [comment]
@tgxworld I checked on the backup code situation. Using the "Disable All" button triggers:

`current_user.user_second_factors.destroy_all`

Because of this, any `current_user.user_second_factors.backup_codes` are inherently destroyed.

One note: when we remove every TOTP method and security key **individually**, the backup codes do remain in database, but the codes are not available to regenerate/disable through the UI until at least one second factor is added. I think this is probably fine, but figured it's worth mentioning the behavior.

**Review by tgxworld — APPROVED** [review]

**Comment by lukastribus** [comment]
> One note: when we remove every TOTP method and security key **individually**, the backup codes do remain in database, but the codes are not available to regenerate/disable through the UI until at least one second factor is added. This has been the behavior prior to this PR and I think it's probably fine, but figured it's worth mentioning it.

@tshenry it is not enough, people continue to get locked out of their accounts due to this trap.

Please fix the broken login code that engages 2FA code paths, even though 2FA is disabled, just because backup codes are present. That is the real solution to this problem, not automatic backup code removal.


Also see:
https://meta.discourse.org/t/admin-locked-out-of-site-after-deleting-two-factor-keys-from-prefs/184193/

**Pull request merged**